README.md
# clojturtle

clojure-turtle fractals


## tree

![tree-fractal](tree.png)

## c-curve

![cline-fractal](cline.png)

## sierpinski
![sierpinski-fractal](sierp.png)

## koch
![koch-curve](koch.png)

## tiles
![tiles](tile.png)

## rings
![rings](rings.png)

## terdragon
![terdragon-curve](terdragon.png)

## pentaplex
![pentaplex](pentaplex.png)

## dragon curve
![dragon-curve](dragon.png)


## Usage

`lein repl`
`(new-window {:size [920 900]})`

### tree
`(setxy 100 -200)`

`(tree/tree [0 100 100] 5 200 true)`

### Lévy C curve
`(setxy -100 -200)`

`(cline 8 300)`

### Sierpiński
`(setxy -400 -400)`

`(sierpi 15 800)`

### Koch curve (regular and quadratic type)
`(clean) (setxy -200 -400) (koch 20 800) (setxy 0 -400) (squarekoch 4 10)`

### Tiles
`(repeat 4 (all (tile 4 10) (right 90)))`

### Rings
`(repeat 4 (all (rings 4 5) (right 90)))`


### Terdragon
`(setxy -635 0) (clean) (terdragon 9 8 [150 -100 110])`

### Pentaplex
`(repeat 5 (all (pentaplex 4 10) (right 36) (right 36)))`

### Dragon curve
` (setxy -435 -250) (clean) (right 180) (dragon 13 4)`

## License


<p xmlns:dct="http://purl.org/dc/terms/" xmlns:vcard="http://www.w3.org/2001/vcard-rdf/3.0#">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0" />
  </a>
  <br />
  To the extent possible under law,
  <span resource="[_:publisher]" rel="dct:publisher">
    <span property="dct:title">Henna Kalliokoski</span></span>
  has waived all copyright and related or neighboring rights to
  this work.
This work is published from:
<span property="vcard:Country" datatype="dct:ISO3166"
      content="FI" about="[_:publisher]">
  Finland</span>.
</p>

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- 

## [0.1.1] - 2019-10-17
### Changed
- Changed license from epl-2.0 to CC0

### Removed
- 

### Fixed
- 



[Unreleased]: 
[0.1.1]: 


test/clojturtle/core_test.clj
(ns clojturtle.core-test
  (:require [clojure.test :refer :all]
            [clojturtle.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clojturtle/tree.clj
(ns clojturtle.tree
  (:require [clojure.set :as set]
            [clojure-turtle.core :refer :all]))

(defn leaf []
  (let [size (rand-int 3)
        angle (get @(right 0) :angle)]
    (do (repeat 10 (all (forward size) (right 8)))
        (right 100)
        (repeat 10 (all (forward size) (right 8)))
        (right 135)
        (forward (* 5 size))
        (back (* 5 size))
        (setheading angle))))

(defn tree [tcolor minl length leafy?]
  (let [r (first tcolor)
        g (second tcolor)
        b (last tcolor)]
    (color tcolor)
    (if (< length minl)
      (do
        (wait 10)
        (forward length)
        (when leafy? (leaf))
        (back length))
      (do (forward length)
          (left 25)
          (tree [(+ 40 r) g (- b 10)] minl (* length 0.8) leafy?)
          (right 75)
          (tree [r (+ g 10) b] minl (* length 0.5) leafy?)
          (left 50)
          (back length)))))

(defn simptree [minl length]
  (if (< length minl)
    (do
      (forward length)
      (back length))
    (do
      (forward length)
      (left 25)
      (simptree minl (* length 0.8))
      (right 75)
      (simptree minl (* length 0.5))
      (left 50)
      (back length))))


src/clojturtle/core.clj
(ns clojturtle.core
  (:require [clojure.set :as set]
            [clojure-turtle.core :refer :all]
            [clojturtle.tree :as tree]))


(defn spiral [minl ang length]
  (if (< length minl)
    nil
    (do (forward length)
        (right ang)
        (wait 10)
        (spiral minl ang (- length minl))
        (left ang)
        (back length))))

(defn koch [minl length]
  (if (< length minl)
    (forward length)
    (do 
      (koch minl (/ length 3.0))
      (left 60)
      (koch minl (/ length 3.0))
      (right 120)
      (koch minl (/ length 3.0))
      (left 60)
      (koch minl (/ length 3.0)))))

(defn squarekoch [dep length]
  (if (zero? dep)
      (forward length)
      (do 
        (squarekoch (dec dep) length)
        (right 90)
        (squarekoch (dec dep) length)
        (left 90)
        (squarekoch (dec dep) length)
        (left 90)
        (squarekoch (dec dep) length)
        (right 90)
        (squarekoch (dec dep) length))))

(defn cline [minl length]
  (if (< length minl)
    (forward length)
    (do
      (left 90)
      (cline minl (/ length 2.0))
      (right 90)
      (cline minl (/ length 2.0))
      (cline minl (/ length 2.0))
      (right 90)
      (cline minl (/ length 2.0))
      (left 90))))

(defn sierpi [minl length]
  (repeat 3
    (all
      (when (> length minl) (sierpi minl (/ length 2.0)))
      (wait 10)
      (forward length)
      (right 120))))

(defn tile [dep length]
  (if (zero? dep)
    (forward length)
    (do
      (tile (dec dep) length)
      (tile (dec dep) length)
      (right 90)
      (tile (dec dep) length)
      (left 90)
      (tile (dec dep) length)
      (right 90)
      (tile (dec dep) length)
      (right 90)
      (tile (dec dep) length)
      (tile (dec dep) length)      
      ))
  )

(defn rings [dep length]
  (if (zero? dep)
    (forward length)
    (do
      (rings (dec dep) length)
      (rings (dec dep) length)
      (right 90)
      (rings (dec dep) length)
      (right 90)
      (rings (dec dep) length)
      (right 90)
      (rings (dec dep) length)
      (right 90)
      (rings (dec dep) length)
      (right 90)
      (rings (dec dep) length)
      (left 90)
      (rings (dec dep) length))))

(defn terdragon [dep length tcolor]
  (let [r (first tcolor)
        g (second tcolor)
        b (last tcolor)]
    (color tcolor)
    (if (zero? dep)
      (forward length)
      (do 
        (terdragon (dec dep) length [(+ 20 r) g (/ b dep)])
        (left 120)
        (terdragon (dec dep) length [r (+ g 40) b])
        (right 120)
        (terdragon (dec dep) length [r (+ 60 g) (+ 60 b)])
        ))))

(defn pentaplex [dep length]
  (if (zero? dep)
    (forward length)
    (do
      (pentaplex (dec dep) length)
      (right 36)
      (right 36)
      (pentaplex (dec dep) length)
      (right 36)
      (right 36)
      (pentaplex (dec dep) length)
      (right 180)
      (pentaplex (dec dep) length)
      (left 36)
      (pentaplex (dec dep) length)
      (right 36)
      (right 36)
      (pentaplex (dec dep) length)      
      )))


(declare dragon2)

(defn dragon [dep length]
  (if (zero? dep)
       (forward length)
       (do
         (dragon (dec dep) length)
         (right 90)
         (dragon2 (dec dep) length)
         (forward length)
         (right 90)
         )))

(defn dragon2 [dep length]
  (if (zero? dep)
    (forward length)
    (do 
      (left 90)
      (forward length)
      (dragon (dec dep) length)
      (left 90)
      (dragon2 (dec dep) length))))

      

(defn -main []
  (new-window {:size [920 900]})
  (setxy 100 -200)
  (tree/tree [50 50 255] 5 200 true)
  (clean)
  (setxy -100 -200)
  (spiral 5 60 300))

